accepted/targeting-packs/targeting-packs.md

```diff
@@ -0,0 +1,156 @@
+Framework doesn't require a Windows machine with a machine-wide installation of
+the corresponding targeting pack.
+
+The .NET Framework supports targeting different versions of the platform
+regardless of the .NET Framework version that is installed on the developer's
+machine. This allows developers to target a lower version than what they use
+themselves. This is especially useful for in-place updates where the older
+version of the framework cannot be installed side-by- side.
+
+That mechanism is called *targeting packs*, which is simply a set of assemblies
+that describe the public APIs of a particular version of the .NET Framework.
+Since those assemblies are only used by the compiler, they don't need to contain
+any code, which we call [reference assemblies](https://www.youtube.com/watch?v=EBpY1UMHDY8).
+
+Unfortunately, the targeting packs aren't available by themselves; they are
+redistributed as part of the .NET Framework Developer Packs which bundle both
+the runtime as well as the targeting pack. This design ensures that developers
+can never be in a state where they have a targeting pack that is higher than the
+.NET Framework version they have installed. That design made sense coming from a
+Visual Studio-first perspective but doesn't serve us well today:
+
+* **It's not cross-platform friendly**. The .NET Framework Developer Packs are
+  only available for Windows which makes the targeting packs indirectly Windows-
+  only too.
+
+* **It's not CI friendly**. Requiring a machine-wide install isn't friendly for
+  cloud-hosted CI machines as developers typically lack permissions to install
+  software.
+
+This document proposes to offer the targeting packs via NuGet that they can be
+acquired in a non-impactful way. It builds on the following existing specs:
+
+* [Phillip Carter](https://github.com/cartermp): [Add Target Packs on NuGet](https://github.com/dotnet/core/pull/64)
+* [Daniel Plaisted](https://github.com/dsplaisted): [Reference Assembly NuGet packages](https://gist.github.com/dsplaisted/83d67bbcff9ec1d0aff1bea1bf4ad79a)
+* [Prototype](https://github.com/dsplaisted/ReferenceAssemblyPackages)
+
+## Scenarios and User Experience
+
+* Build a project which is multi-targeted to .NET Core or .NET Standard as well
+  as .NET Framework on Mac OS or Linux using the .NET CLI
+
+* Build a project targeting Mono (which uses the same Target Framework
+  Identifier as .NET Framework) using the .NET CLI
+
+## Requirements
+
+### Goals
+
+* Doesn't require running an installer
+* Works cross-platform
+* Can be used in SDK-style multi-targeting projects
+* Prefer the centralized targeting pack if it's present. First, this ensures we
+  don't change the performance characteristic of exiting solutions but also
+  allows working around some issue due to build extensions by installing the
+  targeting pack.
+* Supports `<PreserveCompilationContext>` so that ASP.NET Razor compilation
```

Comment:
I've filed https://github.com/dotnet/sdk/issues/2054 to track fixing `PreserveCompilationContext`.  I don't think we should block the rest of the work on this issue though.

Reply:
I'm not sure what you mean "block the rest of the work", but we can't make using the NuGet targeting pack the default in SDK projects without this issue being addressed.  Or else you will break every ASP.NET MVC/Razor Page app running on the desktop .NET Framework.

Reply:
As I understand it, the idea for the first turn of the crank is that you can build without the program files TP but razor will still need it to be there to run. It needn't break anything that already works. You can build against nuget and dynamically compile against the same content in program files at runtime.

Step 2 would be to write out the actual nuget location that build used, and plumb that through for runtime compilation consumption.

Reply:
The current plan is to only reference the NuGet packages if the targeting pack isn't found in program files.  So if you have the reference assemblies installed, everything would continue to work.

I think everything would still work even if we did use the reference assembly packages for compilation as long as you have the reference assemblies installed.  Different parts of the build would be getting the reference assemblies from different locations, but they should be the same files.

Reply:
(I tried to say the same thing, but Daniel worded it better. :))

Reply:
Sounds good.  I just wanted to call out that this area is a bug factory, and often gets overlooked.